### PR TITLE
Engine - Change logging methodology

### PR DESCRIPTION
**Pull request by nmkoremblum**
|Related issue|
|---|
|#16358|

As described in issue #16358, this PR aims to modify the logging code to use the `spdlog` library as the logging library for the engine.

### CODE REVIEW

File: src/engine/test/engineTestMain.cpp [hunk 1]
```diff
@@ -6,7 +6,7 @@
 int main(int argc, char** argv)
 {
     logging::LoggingConfig logConfig;
-    logConfig.logLevel = logging::LogLevel::Off;
+    logConfig.logLevel = "off";
     logging::loggingInit(logConfig);
 
     ::testing::InitGoogleTest(&argc, argv);
```
Comment (juliancnn) [inline src/engine/test/engineTestMain.cpp:12]: What is the use of this file? Wasn't the main test removed?

File: src/engine/test/source/api/catalog/catalog_test.cpp [hunk 2]
```diff
@@ -1,33 +1,49 @@
...
+    virtual void SetUp()
+    {
+        // Logging setup
+        logging::LoggingConfig logConfig;
+        logConfig.logLevel = "off";
+        logConfig.filePath = logging::DEFAULT_TESTS_LOG_PATH;
+        logging::loggingInit(logConfig);
+    }
+
+    virtual void TearDown() {}
```
Comment (juliancnn) [inline src/engine/test/source/api/catalog/catalog_test.cpp:17]: Why are they virtual methods?

File: src/engine/test/source/base/utils/wazuhProtocol/wazuhRequest_test.cpp [hunk 3]
```diff
@@ -4,6 +4,55 @@
...
 #include <logging/logging.hpp>
 
+void initLogging(void)
+{
+    // Logging setup
+    logging::LoggingConfig logConfig;
+    logConfig.logLevel = "off";
+    logConfig.filePath = logging::DEFAULT_TESTS_LOG_PATH;
+    logging::loggingInit(logConfig);
+}
```
Comment (juliancnn) [inline src/engine/test/source/base/utils/wazuhProtocol/wazuhRequest_test.cpp:14]: Can this function be unified for all tests in a single file so as not to duplicate code, something like a base module for the tests?

**Review by juliancnn — CHANGES_REQUESTED** [review]

File: src/engine/source/base/logging/logging.hpp [hunk 4]
```diff
@@ -1,71 +1,103 @@
...
-    Debug = 0,
-    Info,
-    Warn,
-    Error,
-    Off,
-};
 
-// Not great but avoiding having a cpp file for this
+constexpr char DEFAULT_LOG_PATH[] {"/dev/stderr"};
+constexpr char DEFAULT_TESTS_LOG_PATH[] {"/tmp/engine_code_tests.log"};
```
Comment (JavierBejMen) [inline src/engine/source/base/logging/logging.hpp:16]: This should be in the test, not here.

File: src/engine/source/base/logging/logging.hpp [hunk 5]
```diff
@@ -1,71 +1,103 @@
...
+        spdlog::flush_every(std::chrono::milliseconds(cfg.flushInterval));
+
+        // Logger initialization ("default" is the logger name, it can be any custom name)
+        spdlog::basic_logger_mt("default", cfg.filePath, doTruncateFile);
     }
-    else
+    catch (const spdlog::spdlog_ex& ex)
     {
-        fmtlog::setLogFile(stderr, false);
+        std::cerr << "Log initialization failed: " << ex.what() << std::endl;
```
Comment (JavierBejMen) [inline src/engine/source/base/logging/logging.hpp:72]: We should raise an exception, or the one that is already thrown, and manage it on the main or module's run function.

**Review by JavierBejMen — CHANGES_REQUESTED** [review]

File: src/engine/test/engineTestMain.cpp (hunk 1 above)
Comment (nbertoldo) [reply src/engine/test/engineTestMain.cpp]: Yes, I will remove it.

File: src/engine/test/source/base/utils/wazuhProtocol/wazuhRequest_test.cpp (hunk 3 above)
Comment (nbertoldo) [reply src/engine/test/source/base/utils/wazuhProtocol/wazuhRequest_test.cpp]: Done

File: src/engine/source/base/logging/logging.hpp (hunk 4 above)
Comment (nbertoldo) [reply src/engine/source/base/logging/logging.hpp]: Yes, that's right

File: src/engine/source/base/logging/logging.hpp (hunk 5 above)
Comment (nbertoldo) [reply src/engine/source/base/logging/logging.hpp]: Ok, I'll take your suggestion

File: src/engine/source/base/CMakeLists.txt [hunk 6]
```diff
@@ -41,17 +41,18 @@ endif()
...
 CPMAddPackage(
-    NAME fmtlog
-    GITHUB_REPOSITORY MengRao/fmtlog
-    GIT_TAG "v2.1.3"
-    DOWNLOAD_ONLY YES
+    NAME spdlog
+    GITHUB_REPOSITORY gabime/spdlog
+    GIT_TAG "v1.11.0"
+    OPTIONS "SPDLOG_FMT_EXTERNAL_HO ON"
     )
```
Comment (juliancnn) [inline src/engine/source/base/CMakeLists.txt:56]: Fix syntax and If use `SPDLOG_FMT_EXTERNAL_HO` then, should be explicit the ftm HO option:
```
CPMAddPackage(
    NAME fmt
    GITHUB_REPOSITORY fmtlib/fmt
    GIT_TAG "8.1.1"
    OPTIONS "FMT_HEADER_ONLY ON"
)

CPMAddPackage(
    NAME spdlog
    GITHUB_REPOSITORY gabime/spdlog
    GIT_TAG "v1.11.0"
    OPTIONS "SPDLOG_FMT_EXTERNAL_HO ON"
)
```

**Review by juliancnn — CHANGES_REQUESTED** [review]
image

File: src/engine/test/source/base/utils/tests/testsCommon.hpp [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+void inline initLogging(void)
+{
+    // Logging setup
+    logging::LoggingConfig logConfig;
+    logConfig.logLevel = "off";
+    logConfig.filePath = DEFAULT_TESTS_LOG_PATH;
+    logging::loggingInit(logConfig);
+}
+
+#endif // _H_TESTS_COMMON
```
[the remark this answers is not in the document]
Comment (nbertoldo) [reply src/engine/test/source/base/utils/tests/testsCommon.hpp]: Moved to `.../test/source/common/`

File: src/engine/source/base/CMakeLists.txt (hunk 6 above)
Comment (nbertoldo) [reply src/engine/source/base/CMakeLists.txt]: Done

**Review by LucioDonda — APPROVED** [review]
All test passed, from ctest and from the plugin, so it LGTM!

**Review by JavierBejMen — APPROVED** [review]
LGTM

**Pull request merged**